### ISSUE TITLE
HTTP2: Exceptions being thrown while posting multiform Data

### ISSUE THREAD

**Issue by Anipik** [dotnet/corefx#39511]
```
dotnet run -n 10 -maxContentLength 4000 -http 2.0 -ops 7 -numParameters 200
```

```C#
[15-07-2019 16:23:17] Total: 76,700
        POST Multipart Data           Success: 76,680   Fail: 20

Error from iteration 7664 (POST Multipart Data) in task 7 with 76680 successes / 20 fails:
System.Net.Http.HttpRequestException: Error while copying content to a stream.
 ---> System.IO.IOException: The request was aborted.
 ---> System.IO.IOException: The response ended prematurely, with at least 9 additional bytes expected.
   at System.Net.Http.Http2Connection.ReadAtLeastAsync(Stream stream, Memory`1 buffer, Int32 minReadBytes)
   at System.Net.Http.Http2Connection.EnsureIncomingBytesAsync(Int32 minReadBytes)
   at System.Net.Http.Http2Connection.ReadFrameAsync(Boolean initialFrame)
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync()
   --- End of inner exception stack trace ---
   at System.Net.Http.Http2Connection.Http2Stream.CheckIfDisposedOrAborted()
   at System.Net.Http.Http2Connection.Http2Stream.TryReadFromBuffer(Span`1 buffer)
   at System.Net.Http.Http2Connection.Http2Stream.ReadDataAsync(Memory`1 buffer, CancellationToken cancellationToken)
   at System.IO.Stream.CopyToAsyncInternal(Stream destination, Int32 bufferSize, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionResponseContent.SerializeToStreamAsync(Stream stream, TransportContext context, CancellationToken cancellationToken)
   at System.Net.Http.HttpContent.LoadIntoBufferAsyncCore(Task serializeToStreamTask, MemoryStream tempBuffer)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpContent.LoadIntoBufferAsyncCore(Task serializeToStreamTask, MemoryStream tempBuffer)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Program.<>c__DisplayClass1_0.<<Run>b__9>d.MoveNext() in C:\git\corefx\src\System.Net.Http\tests\StressTests\HttpStress\Program.cs:line 269
--- End of stack trace from previous location where exception was thrown ---
   at Program.<>c__DisplayClass1_4.<<Run>b__43>d.MoveNext() in C:\git\corefx\src\System.Net.Http\tests\StressTests\HttpStress\Program.cs:line 634

Error from iteration 7632 (POST Multipart Data) in task 3 with 76680 successes / 20 fails:
System.Net.Http.HttpRequestException: Error while copying content to a stream.
 ---> System.ObjectDisposedException: The object was disposed while operations were in progress.
Object name: 'CreditManager:Http2Connection:_connectionWindow'.
   at System.Net.Http.Http2Connection.SendStreamDataAsync(Int32 streamId, ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.Http2Stream.SendDataAsync(ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.Net.Http.MultipartContent.SerializeToStreamAsync(Stream stream, TransportContext context)
   at System.Net.Http.HttpContent.CopyToAsyncCore(ValueTask copyTask)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpContent.CopyToAsyncCore(ValueTask copyTask)
   at System.Net.Http.Http2Connection.Http2Stream.SendRequestBodyAsync(CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Program.<>c__DisplayClass1_0.<<Run>b__9>d.MoveNext() in C:\git\corefx\src\System.Net.Http\tests\StressTests\HttpStress\Program.cs:line 269
--- End of stack trace from previous location where exception was thrown ---

However, the 1.1 version never fails or throws exception.
```
cc @stephentoub @danmosemsft

**Comment by stephentoub** [comment]
The abort here is being caused by https://github.com/aspnet/AspNetCore/issues/11804, in particular the "Reading is already in progress."  You can see that these exceptions happen on the client in response to the server encountering that error (turn on `-aspnetlog`).

Even when that happens, we really shouldn't be propagating an ObjectDisposedException like this to the caller.  I just tried with the latest in master, and I see similar exceptions, so it's still there.
```
Error from iteration 3076 (POST Multipart Data) in task 9 with 30595 successes / 20 fails:
System.Net.Http.HttpRequestException: Error while copying content to a stream.
 ---> System.ObjectDisposedException: The object was disposed while operations were in progress.
Object name: 'CreditManager:Http2Connection:_connectionWindow'.
   at System.Net.Http.TaskCompletionSourceWithCancellation`1.WaitWithCancellationAsync(CancellationToken cancellationToken) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\TaskCompletionSourceWithCancellation.cs:line 30
   at System.Net.Http.Http2Connection.SendStreamDataAsync(Int32 streamId, ReadOnlyMemory`1 buffer, CancellationToken cancellationToken) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Connection.cs:line 1182
   at System.Net.Http.Http2Connection.Http2Stream.SendDataAsync(ReadOnlyMemory`1 buffer, CancellationToken cancellationToken) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Stream.cs:line 706
   at System.Net.Http.MultipartContent.SerializeToStreamAsyncCore(Stream stream, TransportContext context, CancellationToken cancellationToken) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\MultipartContent.cs:line 195
   at System.Net.Http.HttpContent.CopyToAsyncCore(ValueTask copyTask) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\HttpContent.cs:line 375
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpContent.CopyToAsyncCore(ValueTask copyTask) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\HttpContent.cs:line 375
   at System.Net.Http.Http2Connection.Http2Stream.SendRequestBodyAsync(CancellationToken cancellationToken) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Stream.cs:line 168
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Connection.cs:line 1663
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\HttpConnectionPool.cs:line 541
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\RedirectHandler.cs:line 33
   at System.Net.Http.HttpClient.FinishSendAsyncBuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts) in d:\repos\corefx\src\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 546
   at Program.<>c__DisplayClass1_0.<<Run>b__9>d.MoveNext() in d:\repos\corefx\src\System.Net.Http\tests\StressTests\HttpStress\Program.cs:line 265
--- End of stack trace from previous location where exception was thrown ---
   at Program.<>c__DisplayClass1_4.<<Run>b__41>d.MoveNext() in d:\repos\corefx\src\System.Net.Http\tests\StressTests\HttpStress\Program.cs:line 602
```

**Comment by stephentoub** [comment]
Those object disposed exceptions are coming from here:
https://github.com/dotnet/corefx/blob/8973a72af9e3a2837d18e73041e2c8090d6871d1/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs#L123
triggered by this call stack:
```
   at System.Net.Http.CreditManager.CreateObjectDisposedException(Boolean forActiveWaiter) in D:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\CreditManager.cs:line 129
   at System.Net.Http.CreditManager.Dispose() in D:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\CreditManager.cs:line 123
   at System.Net.Http.Http2Connection.CheckForShutdown() in D:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Connection.cs:line 1423
   at System.Net.Http.Http2Connection.AbortStreams(Exception abortException) in D:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Connection.cs:line 1348
   at System.Net.Http.Http2Connection.Abort(Exception abortException) in D:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Connection.cs:line 1289
   at System.Net.Http.Http2Connection.ProcessIncomingFramesAsync() in D:\repos\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\Http2Connection.cs:line 235
```
so the operation is being aborted, but because it's currently waiting for credit, it's ending up with an ObjectDisposedException.

@geoffkizer, @krwq, what is the failure mode supposed to be here?

**Comment by krwq** [comment]
@stephentoub it feels to me we should not throw ObjectDisposed while calling Abort - Abort tries to get rid of the connection but it's already dead so there is nothing here which should throw I think (perhaps exception set by Abort)

**Comment by stephentoub** [comment]
The exception isn't propagating to the caller of Abort; it's being stored into each of the waiters currently waiting on the CreditManager for credit.  #39520 changes it to instead pass to those waiters the _abortException that's causing things to get torn down, so that this at least doesn't look like a bug and provides some meaningful information.  Subsequently we can see if we can do better as part of all of @geoffkizer's cleanup work.

**Comment by Anipik** [comment]
@stephentoub This error occured again. I am using the lastest preview8 sdk

```C#
error from iteration 90 (POST Duplex Slow) in task 2 with 234 successes / 3 fails:
System.Net.Http.HttpRequestException: Error while copying content to a stream.
 ---> System.ObjectDisposedException: Cannot access a disposed object.
Object name: 'CreditManager:Http2Connection:_connectionWindow'.
   at System.Net.Http.CreditManager.RequestCreditAsync(Int32 amount, CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.SendStreamDataAsync(Int32 streamId, ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.Http2Stream.SendDataAsync(ReadOnlyMemory`1 buffer, CancellationToken cancellationToken)
   at Program.ByteAtATimeNoLengthContent.SerializeToStreamAsync(Stream stream, TransportContext context) in C:\git\corefx\src\System.Net.Http\tests\StressTests\HttpStress\Program.cs:line 839
   at System.Net.Http.HttpContent.CopyToAsyncCore(ValueTask copyTask)
   --- End of inner exception stack trace ---
   at System.Net.Http.HttpContent.CopyToAsyncCore(ValueTask copyTask)
   at System.Net.Http.Http2Connection.Http2Stream.SendRequestBodyAsync(CancellationToken cancellationToken)
   at System.Net.Http.Http2Connection.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken)
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts)
   at Program.RequestContext.SendAsync(HttpRequestMessage request, HttpCompletionOption httpCompletion, Nullable`1 token) in C:\git\corefx\src\System.Net.Http\tests\StressTests\HttpStress\Program.cs:line 748
   at Program.<>c__DisplayClass1_0.<<Run>b__10>d.MoveNext() in C:\git\corefx\src\System.Net.Http\tests\StressTests\HttpStress\Program.cs:line 278
--- End of stack trace from previous location where exception was thrown ---
   at Program.<>c__DisplayClass1_4.<<Run>b__40>d.MoveNext() in C:\git\corefx\src\System.Net.Http\tests\StressTests\HttpStress\Program.cs:line 607
```


// aspnetlog

```c#
fail: Microsoft.AspNetCore.Server.Kestrel[13]
      Connection id "0HLOCEOM5JDDJ", Request id "0HLOCEOM5JDDJ:00000211": An unhandled exception was thrown by the application.
System.IO.IOException: The request stream was aborted.
 ---> Microsoft.AspNetCore.Connections.ConnectionAbortedException: The HTTP/2 connection faulted.
 ---> Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2ConnectionErrorException: HTTP/2 connection error (STREAM_CLOSED): The client sent a DATA frame to closed stream ID 525.
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessDataFrameAsync(ReadOnlySequence`1& payload)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2Connection.ProcessRequestsAsync[TContext](IHttpApplication`1 application)
   --- End of inner exception stack trace ---
   --- End of inner exception stack trace ---
   at System.IO.Pipelines.PipeCompletion.ThrowLatchedException()
   at System.IO.Pipelines.Pipe.GetReadResult(ReadResult& result)
   at System.IO.Pipelines.Pipe.GetReadAsyncResult()
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http2.Http2MessageBody.ReadAsync(CancellationToken cancellationToken)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpRequestStream.ReadAsyncInternal(Memory`1 buffer, CancellationToken cancellationToken)
   at Program.<>c.<<Run>b__1_31>d.MoveNext() in C:\git\corefx\src\System.Net.Http\tests\StressTests\HttpStress\Program.cs:line 492
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.AspNetCore.Routing.EndpointMiddleware.<Invoke>g__AwaitRequestTask|6_0(Endpoint endpoint, Task requestTask, ILogger logger)
   at Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Http.HttpProtocol.ProcessRequests[TContext](IHttpApplication`1 application)
```


Do we need to make a similar change here as well ? https://github.com/dotnet/corefx/blob/3c71b23c676c3538b33f4ae653cf6f85de77e44d/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/CreditManager.cs#L48

**Comment by karelz** [comment]
Fixed in master (3.0) in PR #39520 before final integration into release/3.0 branch.
